Model methods

JaviCerveraIngram opened the pull request:
Added methods to Asset and Fulfillment models, to simplify getting parameters and items by identifier, and listing new, changed and removed items.

Fixed what I think is a bug in ItemSchema, where old_quantity was defined as a string instead of int (please verify this).

ht-albert commented on connect/models/asset.py:
```diff
@@ -15,7 +15,17 @@
+            return [a for a in self.items if a.id == id_][0]
+        except IndexError:
+            return None
 
```
Why not?
```
def _get_by_id(objects, identity):
    try:
        return [obj for obj in objects if obj.id == identity][0]
    except IndexError:
        return None

def get_parameter_by_id(self, identity):
    return self._get_by_id(self.params, identity)

def get_item_by_id(self, identity):
    return self._get_by_id(self.items, identity)
```

ht-albert replied:
And move this method `_get_by_id` to BaseModel

JaviCerveraIngram replied:
Thank you, it's much better like that. I have also used filter() in _get_by_id as you suggested for the Fulfillment methods.

ht-albert replied:
But you no wrote tests :(

JaviCerveraIngram replied:
I'm on it! I'll update the PR soon :)

ht-albert commented on connect/models/fulfillment.py:
```diff
@@ -13,7 +13,17 @@
+    @property
+    def removed_items(self):
+        return [i for i in self.asset.items if i.quantity == 0 and i.old_quantity > 0]
 
```
Use `filter`
`filter(lambda item: item. quantity > 0 and item.old_quantity == 0, self.asset.items)`

JaviCerveraIngram replied:
Done.

JaviCerveraIngram commented:
> Please write tests
> And fix comments

I have added the unit tests. I used a JSON response file from the PHP SDK to do the tests. The type of the fields named 'external_id' on all models was an int, so I changed that type on all the schemas to fields.Int(). If a string is provided, it still parses the model correctly, but the type is converted to an integer. Please confirm that 'external_id' will always have a numeric value (other ids remain as strings).

ht-albert commented:
I checked and public API return `external_id` with type `str`

JaviCerveraIngram commented:
Perfect, I have changed the JSON response file and reverted to the Str type for external_id.

marcserrat commented on connect/models/asset.py:
```diff
@@ -15,7 +15,11 @@
+    def get_param_by_id(self, identity):
+        return self._get_by_id(self.params, identity)
+
+    def get_item_by_id(self, identity):
```
This is deprecated......don't add it or leave there for providers. Please create get_item_by_mpn. That's how vendors works, ids are autogenerated since v13 and in phpruntime this is due backwards compatibility only, if you check documentation all goes by mpn

JaviCerveraIngram replied:
I'll make the PR update soon with get_item_by_mpn replacing get_item_by_id. I understand that get_param_by_id is not deprecated?

marcserrat replied:
FYI: get_item_by_id has no business sense for an ISV, they don't know the ids since they don't introduce them in connect, they introduce MPNs and this is what's known by vendor.

JaviCerveraIngram replied:
Makes sense. Thank you!
